fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+            ));
+        }
+
+        let mut data = BTreeMap::new();
+
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_file()
+                && path
+                    .extension()
+                    .map_or(false, |ext| ext == "json")
+            {
+                let key: K = self.extract_key_from_file_path(&path)?;
+                let file = File::open(&path)?;
+                let value: V =
+                    serde_json::from_reader(file).map_err(|err| {
+                        ArklibError::Storage(
+                            self.label.clone(),
+                            err.to_string(),
+                        )
+                    })?;
+
+                data.insert(key.clone(), value);
+
+                if let Ok(metadata) = fs::metadata(&path) {
+                    if let Ok(modified) = metadata.modified() {
+                        self.written_to_disk.insert(key.clone(), modified);
+                        self.modified.insert(key, modified);
+                    }
+                }
+            }
+        }
+
+        self.data = data;
+        Ok(())
+    }
+    /// Resolve differences between memory and disk data
+    fn resolve_divergence(&mut self) -> Result<()> {
+        let new_data = FolderStorage::new("new_data".into(), &self.path)?;
+
+        for (key, new_value) in new_data.data.iter() {
+            if let Some(existing_value) = self.data.get(key) {
+                let existing_value_updated = self
+                    .modified
+                    .get(key)
+                    .and_then(|ram_stamp| {
+                        self.written_to_disk
+                            .get(key)
+                            .map(|disk_stamp| ram_stamp > disk_stamp)
+                    })
+                    .unwrap_or(false);
+
+                // Use monoid to combine value for the given key
+                // if the memory and disk have diverged
+                if existing_value_updated {
+                    let resolved_value = V::combine(existing_value, new_value);
+                    self.data.insert(key.clone(), resolved_value);
+                } else {
+                    self.data.insert(key.clone(), new_value.clone());
+                }
+            } else {
+                self.data.insert(key.clone(), new_value.clone());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remove files from disk that are not present in memory
+    fn remove_files_not_in_ram(&mut self) -> Result<()> {
+        for key in self.deleted_keys.iter() {
+            let file_path = self.path.join(format!("{}.json", key));
+            if file_path.exists() {
+                fs::remove_file(&file_path).expect("Failed to delete file");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn extract_key_from_file_path(&self, path: &Path) -> Result<K> {
```

Comment:
This method doesn't really depend on any fields or properties of `FolderStorage`, so it shouldn't be part of that struct. It would be better as a helper function, and it probably shouldn't be public.